src/colmap/estimators/homography_matrix_test.cc

```diff
@@ -38,38 +38,114 @@
 namespace colmap {
 namespace {
 
-TEST(HomographyMatrix, Estimate) {
+TEST(HomographyMatrix, Minimal) {
+  constexpr size_t kNumPoints = 4;
   for (int x = 0; x < 10; ++x) {
-    Eigen::Matrix3d H0;
-    H0 << x, 0.2, 0.3, 30, 0.2, 0.1, 0.3, 20, 1;
+    Eigen::Matrix3d expected_H;
+    expected_H << x, 0.2, 0.3, 30, 0.2, 0.1, 0.3, 20, 1;
 
     std::vector<Eigen::Vector2d> src;
-    src.emplace_back(x, 0);
-    src.emplace_back(1, 0);
-    src.emplace_back(2, 1);
-    src.emplace_back(10, 30);
-
     std::vector<Eigen::Vector2d> dst;
+    for (size_t i = 0; i < kNumPoints; ++i) {
+      src.push_back(Eigen::Vector2d::Random());
+      dst.push_back((expected_H * src[i].homogeneous()).hnormalized());
+    }
+
+    HomographyMatrixEstimator estimator;
+    std::vector<Eigen::Matrix3d> models;
+    estimator.Estimate(src, dst, &models);
+
+    ASSERT_EQ(models.size(), 1);
+
+    std::vector<double> residuals;
+    estimator.Residuals(src, dst, models[0], &residuals);
 
     for (size_t i = 0; i < 4; ++i) {
-      const Eigen::Vector3d dsth = H0 * src[i].homogeneous();
-      dst.push_back(dsth.hnormalized());
+      EXPECT_LT(residuals[i], 1e-6);
     }
+  }
+}
 
-    HomographyMatrixEstimator est_tform;
+TEST(HomographyMatrix, NonMinimal) {
+  constexpr size_t kNumPoints = 20;
+  for (int x = 0; x < 10; ++x) {
+    Eigen::Matrix3d expected_H;
+    expected_H << x, 0.2, 0.3, 30, 0.2, 0.1, 0.3, 20, 1;
+
+    std::vector<Eigen::Vector2d> src;
+    std::vector<Eigen::Vector2d> dst;
+    for (size_t i = 0; i < kNumPoints; ++i) {
+      src.push_back(Eigen::Vector2d::Random());
+      dst.push_back((expected_H * src[i].homogeneous()).hnormalized());
+    }
+
+    HomographyMatrixEstimator estimator;
     std::vector<Eigen::Matrix3d> models;
-    est_tform.Estimate(src, dst, &models);
+    estimator.Estimate(src, dst, &models);
+    ASSERT_EQ(models.size(), 1);
+
+    std::vector<double> residuals;
+    estimator.Residuals(src, dst, models[0], &residuals);
 
+    for (size_t i = 0; i < 4; ++i) {
+      EXPECT_LT(residuals[i], 1e-6);
+    }
+  }
+}
+
+// Test numerical stability with large coordinates. This is to ensure that the
+// homography matrix estimator is numerically stable despite not using
+// coordinate normalization. We can do this because of double precision.
+TEST(HomographyMatrix, NumericalStability) {
```

Comment:
Tbh I am not sure about the validity of this test. To be able to test numerical stability should we add a small noise to ``(expected_H * x.homogeneous()).hnormalized()`` and relax the checking threshold? 

Also, I personally dont see a clear reason not to have coordinate normalization, since it is generally not the performance bottleneck, or is it?

Reply:
It's less about noise but the fact that the different columns in the constraint matrix will have largely different magnitudes of values, because of some columns set to constants, some of them to a single coordinate and some of them to the product of two coordinates. When the coordinates are large, they will get huge compared to others. I will add another test with adding some noise as you suggest.

Reply:
Added a bunch more tests and also parameterized them to run all of them on different number of points.